### PR TITLE
Fix assertion failure and crash on HTTP header error, BTS-792 (#15728)

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Backport of https://github.com/arangodb/arangodb/pull/15728/

This PR strives to solve two different problems which can occur if an error is detected in the HTTP header. One is an assertion failure in maintainer mode, the other is a proper crash. The latter is particularly dangerous, since it can be used to crash a server in production with a carefully crafted HTTP header.

First the assertion failure:

The `_response` member of the HttpCommTask object was reset in the `on_message_began` callback. This is not reached if an error is detected in the header before actually reading the message. In this case an assertion failure can be triggered when sending the 505 response. The fix is to additionall reset `_response`  in the `on_header_complete` method.

Now the proper crash:

Since we use the llhttp library for parsing HTTP requests, this problem seems to be present. The problem is that in the case of a bad header, we have returned `HPE_OK` which lead to the fact that both an error response was sent out via `sendSImpleResponse` (actually happening in an IO thread asynchronously), and we continued to work on the request and in the end sent out another, potentially positive response.

The fix is to return `HPE_USER` to indicate a header parsing error. But then we also have to react to the fact that llhttp does not hand this error through, but returns HPE_CB_HEADERS_COMPLETE instead. So in this case, we must not close the connection right away, but must first let the response be sent, and then let the connection be closed due to the error.

Fix was validated using HTTP fuzzing functionality soon to be added to devel (https://github.com/arangodb/arangodb/pull/15828).

- [x] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [x] Backports
  - [x] Backport for 3.9: this PR
  - [ ] Backport for 3.8: *(Please link PR)*
  - [ ] Backport for 3.7: *(Please link PR)*

#### Related Information

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

**Review by apetenchea — APPROVED** [review]

File: arangod/GeneralServer/HttpCommTask.cpp [hunk 1]
```diff
@@ -150,12 +151,12 @@ int HttpCommTask<T>::on_header_complete(llhttp_t* p) {
       (p->http_major != 1 || p->http_minor != 1)) {
     me->sendSimpleResponse(rest::ResponseCode::HTTP_VERSION_NOT_SUPPORTED,
                            rest::ContentType::UNSET, 1, VPackBuffer<uint8_t>());
-    return HPE_OK;
+    return HPE_USER;
```
Comment (markuspf) [inline arangod/GeneralServer/HttpCommTask.cpp:154]: What is the error code `HPE_USER` used for in arangodb? (what is the significance of this change?)

Comment (jsteemann) [reply arangod/GeneralServer/HttpCommTask.cpp:154]: `HPE_USER` is not directly used by arangod, but there are several places that compare the return code of the http processing callbacks with `HPE_OK` (e.g. `if (err != HPE_OK) ...`.
Previously we returned `HPE_OK` from here, but the `sendSimpleResponse` already triggered the sending of a response in the background asynchronously. The `HPE_OK` made us continue reading data from the connection and continuing the parsing. Thus we had a race here, and that could actually lead to a crash. 
Returning `HPE_USER` here means: ok, this is not the same as `HPE_OK`, and it is an error that will abort further processing. But it is not any of the predefined errors provided by llhttp. It is basically a catch-all "custom error" happened (in this case: "unsupported HTTP version" or "request entity too large", for which llhttp doesn't have its own error enum values). The fix was validated with an HTTP fuzzer against 3.9.

**Review by neunhoef — APPROVED** [review]
LGTM

**Review by markuspf — APPROVED** [review]
LGTM

File: arangod/GeneralServer/HttpCommTask.cpp (hunk 1 above)
Comment (markuspf) [reply arangod/GeneralServer/HttpCommTask.cpp:154]: :+1:

**Pull request merged**